src/Adapter/ContainerBuilder.php

```diff
@@ -38,35 +47,183 @@
+            $container = new $this->containerClassName();
         }
 
+        return $container;
+    }
+
+    /**
+     * @return SfContainerBuilder
+     *
+     * @throws \Exception
+     */
+    private function compileContainer()
+    {
         $container = new SfContainerBuilder();
+
         $container->addCompilerPass(new LegacyCompilerPass());
-        $loader = new YamlFileLoader($container, new FileLocator(__DIR__));
-        $env = $isDebug ? 'dev' : 'prod';
-        $servicesPath = _PS_CONFIG_DIR_ . "services/${name}/services_${env}.yml";
-        $loader->load($servicesPath);
+        $container->addCompilerPass(new LoadServicesFromModulesPass($this->containerName), PassConfig::TYPE_BEFORE_OPTIMIZATION, 1);
+        $this->loadServices($container);
+
+        //Build extensions
+        $builderExtensions = [
+            new DoctrineBuilderExtension($this->environment),
+        ];
+        /** @var ContainerBuilderExtensionInterface $builderExtension */
+        foreach ($builderExtensions as $builderExtension) {
+            $builderExtension->build($container);
+        }
+
         $container->compile();
 
+        //Dump the container file
         $dumper = new PhpDumper($container);
-        file_put_contents($file, $dumper->dump(array('class' => $containerName)));
+        $this->containerConfigCache->write(
+            $dumper->dump(['class' => $this->containerClassName]),
+            $container->getResources()
+        );
 
         return $container;
     }
+
+    /**
+     * In symfony context doctrine classes (like Table, Entity, ...) are available thanks to
+     * the autoloader. In this specific context we don't have the general autoloader, so we need
+     * to include these classes manually. This is performed in Doctrine\ORM\Configuration::newDefaultAnnotationDriver
+     * which is called in Setup::createAnnotationMetadataConfiguration.
+     */
+    private function loadDoctrineAnnotationMetadata()
+    {
+        Setup::createAnnotationMetadataConfiguration([]);
+    }
+
+    /**
+     * @param SfContainerBuilder $container
+     *
+     * @throws \Exception
+     */
+    private function loadServices(SfContainerBuilder $container)
+    {
+        $loader = new YamlFileLoader($container, new FileLocator(__DIR__));
+        $servicesPath = sprintf(
+            '%sservices/%s/services_%s.yml',
+                _PS_CONFIG_DIR_,
+            $this->containerName,
+            $this->environment->getName()
+        );
+        $loader->load($servicesPath);
+    }
+
+    /**
+     * Loops through all active modules and automatically include their autoload (if present).
+     * Needs to be done as earlier as possible in application lifecycle.
+     *
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    private function loadModulesAutoloader()
+    {
+        $moduleRepository = ModuleRepositoryFactory::getInstance()->getRepository();
+        if (null !== $moduleRepository) {
```

Comment:
return fast instead long condition ```suggestion
        if (null === $moduleRepository) {
            return;
        }
```

Reply:
nice catch!